### ISSUE TITLE
random failing of `t/LocalDB/Index/Blast.t`

### ISSUE THREAD

**Issue by carandraug** [bioperl/bioperl-live#281]
The travis builds are often failing on the test `t/LocalDB/Index/Blast.t`.  The failure is fairly frequently (more than half the times) but not fully reproducible. I have seen it happening on 5.16, 5.18, and 5.20.  The error is always the same though:

```
------------- EXCEPTION: Bio::Root::Exception -------------
MSG: This index file is from version [0.2] - You need to rebuild it to use module version [1.007002]
STACK: Error::throw
STACK: Bio::Root::Root::throw Bio/Root/Root.pm:447
STACK: Bio::Index::Abstract::_type_and_version Bio/Index/Abstract.pm:493
STACK: Bio::Index::Abstract::open_dbm Bio/Index/Abstract.pm:415
STACK: Bio::Index::Abstract::new Bio/Index/Abstract.pm:162
STACK: Bio::Index::Blast::new Bio/Index/Blast.pm:143
STACK: t/LocalDB/Index/Blast.t:20
-----------------------------------------------------------
```

I am unable to reproduce it on my own system.

**Comment by carandraug** [comment]
Seems to be related to the `_version` property of `Bio::Index` instances. Their value is meant some sort of ABI version number for the Index files provided. While some have their own 0.1, 0.2 numbers, others seem to use the version from Bio::Root.  How this versioning is actually causing an error I'm not sure.

**Comment by carandraug** [comment]
I found the source of this.  The issue is that bioperl tests create temporary files in the current working directory without check for those filenames. While the tests do remove the files at the end, when they run in parallel they sometimes clash and there's two tests writing and reading from the same file.

In this specific case, both `t/LocalDB/Index/Index.t` and `t/LocalDB/Index/Blast.t` create a `Wibbl` file. In one case, it's an index for a Fasta file and the other for a Blast file. Both were calling the constructor and because the file already exists, it opens it. The Fasta index uses its own version numbers (version 0.2 currently) while the Blast index uses the Bioperl version (1.007002), which explains the error.

I have now fixed this by ensuring that the Blast Index temporary files are created in a temporary directory with a unique name to avoid clashes (but after looking at the rest of the tests, I'm surprised this is not more common).

However, there's a couple of issues which are real issues in the `Bio::Index` classes:

1. the constructor should have an option to create a file. Actually, it should probably create a file by default, and use a separate method to open an existing file. having the user checking for the presence of the file before the constructor is not good enough since it can still lead to race conditions.

2. the index should check for the type first and then the version. This would have provided a much obvious error message. yes, the version number is different but that really only makes sense when talking about the version of the same thing.
